File: number_of_islands.py
class Solution:
    def numIslands(self, grid: list[list[str]]) -> int:
        def dfs(row, col, grid):
            
            if row < 0 or col < 0 or row >= len(grid) or \
            col >= len(grid[0]) or grid[row][col] != '1':
                return
            
            #mark as visited
            grid[row][col] = 'visited'
            
            for dx, dy in [(-1,0),(1,0),(0,1),(0,-1)]:
                dfs(row + dx, col + dy, grid)
                
        ans = 0

        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == '1':
                    dfs(row, col, grid)
                    ans += 1
        return ans

File: NumberOfIslands.cpp
class Solution {
public:
    int ans=0;
    void dfs(vector<vector<char>>& grid,int i,int j){
       if(i<0 || j<0 || i>=grid.size() || j>=grid[0].size() || grid[i][j]=='0')
       return;
       grid[i][j]='0';
       dfs(grid,i+1,j);
       dfs(grid,i,j+1);
       dfs(grid,i-1,j);
       dfs(grid,i,j-1);
       return;
    }
    int numIslands(vector<vector<char>>& grid) {
        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[i].size();j++){
                if(grid[i][j]=='1')
                ans++,dfs(grid,i,j);
            }
        }
        return ans;
    }
};
